torch/csrc/jit/script/compiler.cpp

```diff
@@ -1364,43 +1364,84 @@ struct to_ir {
-                        ->asValue(lhs.range(), method);
-    if (lhsValue->type()->isSubtypeOf(DynamicType::get())) {
-      // for tensors, emit the corresponding in-place op
-      Symbol op;
+  // Get the appropriate builtin op for this augmented assignment
+  // If the RHS is a tensor, return the corresponding ATen in-place op
+  // If it's a list of scalars, then return the corresponding list augment op
+  Symbol getAugOp(const AugAssign& stmt, bool isTensor) {
       switch (stmt.aug_op()) {
         case '+':
-          op = aten::add_;
-          break;
+          return isTensor ? aten::add_ : aten::_aug_item_add;
         case '-':
-          op = aten::sub_;
-          break;
+          return isTensor ? aten::sub_ : aten::_aug_item_sub;
         case '/':
-          op = aten::div_;
-          break;
+          return isTensor ? aten::div_ : aten::_aug_item_div;
         case '*':
-          op = aten::mul_;
-          break;
+          return isTensor ? aten::mul_ : aten::_aug_item_mul;
         default:
-          throw ErrorReport(stmt) << "Unknown augmented assignment to Tensor: "
+          throw ErrorReport(stmt) << "Unknown augmented assignment: "
                                   << kindToString(stmt.aug_op());
       }
+  }
+
+  // Emit nodes for augmented assignments like `+=`
+  void emitAugAssignment(const AugAssign& stmt) {
+    const auto lhsValue = emitExpr(stmt.lhs());
+    if (lhsValue->type()->isSubtypeOf(DynamicType::get())) {
+      // for tensors, emit the corresponding in-place op
 
-      auto rhs = NamedValue(stmt.rhs().range(), emitExpr(stmt.rhs()));
-      auto self = NamedValue(lhs.range(), lhs.name().name(), lhsValue);
-      auto output = emitBuiltinCall(
+      const auto rhs = NamedValue(stmt.rhs().range(), emitExpr(stmt.rhs()));
+      const auto self = NamedValue(stmt.lhs().range(), "self", lhsValue);
+      const auto output = emitBuiltinCall(
           stmt.range(),
           *method.graph(),
-          op,
+          getAugOp(stmt, /*isTensor=*/true),
           self,
           {rhs},
           {},
           /*required=*/true);
-      environment_stack->setVar(lhs.range(), lhs.name().name(), output);
+
+      if (stmt.lhs().kind() == TK_VAR) {
+        // if the lhs was a simple tensor variable, bind a new version of that
+        // variable to the op output.
+        const auto lhs = Var(stmt.lhs()).name();
+        environment_stack->setVar(lhs.range(), lhs.name(), output);
+      }
+    } else if (stmt.lhs().kind() == TK_SUBSCRIPT) {
+      // If it's a subscript expr that doesn't ultimately evaluate to tensor,
+      // it's a list of primitive types.
+
+      // Process the base list value
+      const auto lhs = Subscript(stmt.lhs());
+      const auto listExpr = lhs.value();
+      const auto listValue = emitExpr(listExpr);
+      const auto listType = listValue->type()->cast<ListType>();
+      JIT_ASSERT(listType != nullptr);
+      JIT_ASSERT(
```

Comment:
this is just based on what we support in the frontend today. It would need to change if we gave users a way of defining nested lists or whatever.

Reply:
Oh, I missed something here because I thought the assert was suppose to be an error check. There is much more major bug in the way this is structured.
You are calling emitExpr on the LHS index twice (once on the first line of the function where the entire LHS is emitted, and again for listValue). This will be wrong:
```
def foo():
   print "HI"

a[foo()] += 4
```

The above will print "HI" twice.